src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1490 @@
+            if (warning2.length < warning1.length)
+                return false;
+            for (int index = 0; index < warning1.length; index++) {
+                if (warning2[index].getPreferredPosition() != warning1[index].getPreferredPosition())
+                    return false;
+            }
+            return true;
+        };
+
+        // Stack traces are ordered top to bottom, and so duplicates always have the same first element(s).
+        // Sort the stack traces lexicographically, so that duplicates immediately follow what they duplicate.
+        Comparator<DiagnosticPosition[]> ordering = (warning1, warning2) -> {
+            for (int index1 = 0, index2 = 0; true; index1++, index2++) {
+                boolean end1 = index1 >= warning1.length;
+                boolean end2 = index2 >= warning2.length;
+                if (end1 && end2)
+                    return 0;
+                if (end1)
+                    return -1;
+                if (end2)
+                    return 1;
+                int posn1 = warning1[index1].getPreferredPosition();
+                int posn2 = warning2[index2].getPreferredPosition();
+                int diff = Integer.compare(posn1, posn2);
+                if (diff != 0)
+                    return diff;
+            }
+        };
+        warningList.sort(ordering);
+
+        // Now emit the warnings, but skipping over duplicates as we go through the list
+        DiagnosticPosition[] previous = null;
+        for (DiagnosticPosition[] warning : warningList) {
+
+            // Skip duplicates
+            if (previous != null && extendsAsPrefix.test(previous, warning))
+                continue;
+            previous = warning;
+
+            // Emit warnings showing the entire stack trace
+            JCDiagnostic.Warning key = Warnings.PossibleThisEscape;
+            int remain = warning.length;
+            do {
+                DiagnosticPosition pos = warning[--remain];
+                log.warning(Lint.LintCategory.THIS_ESCAPE, pos, key);
+                key = Warnings.PossibleThisEscapeLocation;
+            } while (remain > 0);
+        }
+        warningList.clear();
+    }
+
+    // Analyze statements, but stop at (and record) the first warning generated
+    private void analyzeStatements(List<JCStatement> stats) {
+        for (JCStatement stat : stats) {
+            scan(stat);
+            if (copyPendingWarning())
+                break;
+        }
+    }
+
+    @Override
+    public void scan(JCTree tree) {
+
+        // Check node
+        if (tree == null || tree.type == Type.stuckType)
+            return;
+
+        // Sanity check
+        Assert.check(checkInvariants(true, false));
+
+        // Can this expression node possibly leave a 'this' reference on the stack?
+        boolean referenceExpressionNode;
+        switch (tree.getTag()) {
+        case SWITCH_EXPRESSION:
```

Comment:
Missing indent?

Reply:
Thanks - will fix (habits are hard to break).